Automatically geocode on Shelter save

tdooner opened the pull request:
Rather than requiring a backfill rake script, let's just geocode each
Shelter entity before it is saved.

This commit adds the recommended `after_validation :geocode` hook in the
`geocode` gem's documentation.

tdooner committed:
Automatically geocode on Shelter save

Rather than requiring a backfill rake script, let's just geocode each
Shelter entity before it is saved.

This commit adds the recommended `after_validation :geocode` hook in the
`geocode` gem's documentation.

jwo commented:
@tdooner Thanks for your contribution!

We have some bad-data in at least the harvey instance.... The `address` field should probably be combined with the city field to help out with #25  (more info on #28)

Can you add a test for something on "300 Main St" and that it wouldn't geocode without a city? I know that's a bit generic, but I think it's needed.

Also, we should do the same on Needs and CharitableOrganizations

tdooner committed:
Skip geocoding shelters without a city

Apparently there is some bad data in the database for which these
geocodings might fail.

tdooner committed:
Only geocode if the address and city are present and changed

This is an effort to avoid re-breaking
https://github.com/sketch-city/harvey-api/pull/25 and
https://github.com/sketch-city/harvey-api/issues/28.

tdooner commented:
Hey @jwo, can you take another look? Hopefully this will not re-break everything?

jwo commented:
@tdooner I have reservations here; I'll do some testing this weekend to try to see that it won't put pins back in new york city.

Eventually we'll move to geo-coding in the browser via a form (#28). I might feel more comfortable if we just do that.

tdooner commented:
Were the pins in NY because of an ambiguous city name?

I'm going to stop working on this branch then, if you think this is the wrong approach. I thought it might be a good quick fix but apparently not so. Let me know if there is a different approach you would prefer me to spend time investing in.

ironcoconut reviewed (commented):
We really need to add a state and zip code field. That is the best way to ensure the geocoding is correct.

jwo commented:
We'll move forward with browser based geocoding.
